submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,132 @@
+This could be achieved using different cable driver implementations or the same cable driver implementation with
+differing encryption configurations.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with an homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectivityPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectivityPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectivityPoliciesSpec struct {
+    Policies []ClusterConnectivityPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectivityPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    CableDriver *CableDriverSpec `json:"cableDriver,omitempty"`
+}
+
+type CableDriverSpec struct {
+    // Name of the cable driver implementation.
+    Name string `json:"name"`
+
+    // Options specifies key/value configuration parameters for the cable driver.
+    Options map[string]string `json:"options,omitempty"`
+}
+```
+
+The `ClusterConnectivityPolicies` CRD defines a list of policies that are applied in order of precedence. To resolve the
+connectivity policy for a remote cluster's `Endpoint`, Submariner would iterate the list of policies to search for a policy
+whose left and right label selectors match the corresponding `Cluster` resources of the local and remote `Endpoints`. The
+first matching policy is used. If no matching policy is found, the clusters are not connected.
+
+For full mesh connectivity, a policy with empty left and right label selectors would match all clusters. This would be the
+default.
+
+For each connectivity policy, the cable driver and associated configuration options can optionally be specified. If none
+is specified, the default cable driver is used.
+
+The `ClusterConnectivityPolicies` resource would be a singleton with a well-known name and would be maintained on the
```

Comment:
> Not sure I understand what this means, why does it need to be a singleton with a constant name?
> It seems to be quite unlike anything else in K8s..

The `ClusterConnectivityPolicies` defines all the policies in precedent order so there only needs to be one in a cluster set.

Reply:
Why then not just allow multiple such policies to exist?

If multiple policies match the same two clusters (or endpoints?) we can then issue a warning and select arbitrarily one of them (say based on creation date).

Reply:
BTW there is precedence for resources with a well-known name: the submariner `Cluster` and the proposed `ClusterClaim`.

Reply:
> Why then not just allow multiple such policies to exist?
> 
> If multiple policies match the same two clusters (or endpoints?) we can then issue a warning and select arbitrarily one of them (say based on creation date).

We could - that was the original proposal. I'll reverse your question - why not do it this way to avoid those complexities?